### ISSUE TITLE
Unhandled exceptions are being logged before custom IExceptionHandler is being called

### ISSUE THREAD

**Issue by AlOnestone01** [dotnet/aspnetcore#54554]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

When using a custom `IExceptionHandler` an unhandled exception is being logged (e.g. to console) before the custom exception handler is being called. The exception is being logged with the message:

`An unhandled exception has occurred while executing the request`

This happens in `ExceptionHandlerMiddlewareImpl.cs` in the method `private async Task HandleException(HttpContext context, ExceptionDispatchInfo edi)`: `DiagnosticsTelemetry.ReportUnhandledException(_logger, context, edi.SourceException);`

After that the custom `IExceptionHandler`  is being called.

### Expected Behavior

If there is a custom `IExceptionHandler` registered, it is being called without logging anything before. In the `IExceptionHandler` the developer can decide if and how to log the unhandled exception.

### Steps To Reproduce

Minimalistic project for reproduction: https://github.com/AlOnestone01/custom-exception-handling

1. Implement custom `IExceptionHandler`
2. Register the handler: `builder.Services.AddExceptionHandler<GlobalExceptionHandler>();`
3. Register the middleware: app.UseExceptionHandler();
4. Implement an endpoint that throws an exception

### .NET Version

8.0.202

### Anything else?

Sample output of the minimalistic project: 
[exception-log.txt](https://github.com/dotnet/aspnetcore/files/14615890/exception-log.txt)

dotnet --info:
.NET SDK:
 Version:           8.0.202
 Commit:            25674bb2f4
 Workload version:  8.0.200-manifests.8cf8de6d

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19045
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\8.0.202\

.NET workloads installed:
There are no installed workloads to display.

Host:
  Version:      8.0.3
  Architecture: x64
  Commit:       9f4b1f5d66

.NET SDKs installed:
  8.0.202 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.28 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.28 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.17 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 6.0.28 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.17 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found

**Comment by OskarKlintrot** [comment]
This is a confusing behaviour giving the [example in the docs](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/error-handling?view=aspnetcore-8.0#iexceptionhandler). In that example implementation of `IExceptionHandler` logs the exception, which in reality will result in logging the exception twice.

**Comment by captainsafia** [comment]
@AlOnestone01 Thanks for reporting this issue!

This change in behavior is a result of some work we did in .NET 8 to add support for metrics to ASP.NET Core. The docs that @OskarKlintrot references in his comments haven't been updated to reflect this change in behavior.

@JamesNK Assuming this behavior change is intentional, did we author a breaking change notice here?

**Comment by JamesNK** [comment]
Is this a change in behavior? I looked at the PR that added `DiagnosticsTelemetry.ReportUnhandledException` method. The exception was already logged.

https://github.com/dotnet/aspnetcore/pull/46834/files#diff-307cb43e0f10bb8837662fc645a7e0cacaefb6560bdba0cdc8e9648bed392fac

**Comment by emyklebost** [comment]
Just to add to this issue: [A Diagnostic Event will always be sent by the default ExceptionHandler](https://github.com/dotnet/aspnetcore/blob/091da452aab1838fd30f6de1decd921326736146/src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerMiddlewareImpl.cs#L205-L213), regardless of if it is "handled" by an `IExceptionHandler` or not, this will result in the Exception showing up as an unhandled exception in e.g. Application Insights. Don't know if this is intended behavior or not, but something to address if the implementation is changed to not log when a `IExceptionHandler` was able to handle it.

**Comment by axylophon** [comment]
I can confirm the problem. I disabled the Log for Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware to mitigate the log issue. The IExceptionHandler should control the logging.


The Application Insights topic is also an issue. Depending on the IExceptionHandler outcome the exception should be added to Application Insights instead for all possible exceptions.

**Comment by JamesNK** [comment]
It’s too late to do it in 9.

We can look at changing it in 10.

**Comment by andy1547** [comment]
Note if using Serilog rather than Microsoft logger, the configuration for disabling is:
  
```
"Serilog": {
    "MinimumLevel": {
      "Default": "Information",
      "Override": {
        // disable default exception logging
        "Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware": "Fatal",
      }
    }
```
In our case, we wanted to treat some thrown exceptions as client errors and others as server errors, returning 400-499 / 500-599 status codes respectively. Disabling in in-built logging to only log server errors inside of IExceptionHandler was necessary to avoid spamming the logs.

One could take the stance that all exceptions should be logged and instead IActionResult should be explicitly returned for client errors. Our stack made use of various service layers (that have no knowledge of ASP.NET Core) that throw validation errors (that should be treated as client errors). Duplicating this validation logic was infeasible, aswell as adding a try/catch around each of these calls. 
The other approach was introducing an IExceptionFilter (ran for just MVC pipeline) for handling client thrown exceptions, returning an IActionResult. However we decided against this approach as using both IExceptionFilter and a IExceptionHandler together added complexity (we prefer to have all exception handling done in a single place).

Funnily we actually had the reverse issue of not logging explicitly returned ObjectResult 500-599 status codes. The workaround was adding logging in an IAlwaysRunResultFilter.

**Comment by SamJ26** [comment]
When can we expect this update to be released?

**Comment by martincostello** [comment]
If it lands in .NET 10, then November.

**Comment by voroninp** [comment]
@andy1547 , same issue. 

Some exceptions are transformed to status and problem details response for the client and should not be logged. At least not with the `Error` level.
Whereas unforeseen and thus unhandled errors should be logged.

Unfortunately, disabling logs with the override `"Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware": "Fatal"` disables the logging of all exceptions.

Such behavior of the middleware is extremely annoying.

The dirty and not breaking fix could be the check for the presence of some documented flag in exception's metadata:
```
public static void ReportUnhandledException(ILogger logger, HttpContext context, Exception ex)
{
    if (!ex.Data.Contains(DO_NOT_LOG_AS_UNHANDLED))
    {
        logger.UnhandledException(ex);
    }

    if (context.Features.Get<IHttpMetricsTagsFeature>() is { } tagsFeature)
    {
        // Multiple exception middleware could be registered that have already added the tag.
        // We don't want to add a duplicate tag here because that breaks some metrics systems.
        tagsFeature.TryAddTag("error.type", ex.GetType().FullName);
    }
}
```

**Comment by waynebrantley** [comment]
@JamesNK Before it becomes to late to do something about this in dotnet 10, can this be addressed?

**Comment by JamesNK** [comment]
Yeah I plan to do this in 10. I'm really busy right now, but I'll improve this in the second half of the year.

**Comment by BrunoBlanes** [comment]
Quick question, once this is fixed, in case `TryHandleAsync` returns `false`, is it possible for it to finish the HTTP response pipeline then throw?
```csharp
    public async ValueTask<bool> TryHandleAsync(
        HttpContext httpContext,
        Exception exception,
        CancellationToken cancellationToken)
    {
        ProblemHttpResult problem = TypedResults.Problem(
            "Check your logs for more information.",
            statusCode: StatusCodes.Status500InternalServerError);
        httpContext.Response.ContentType = MediaTypeNames.Application.ProblemJson;
        await httpContext.Response.WriteAsJsonAsync(problem.ProblemDetails, cancellationToken);
        return false;
    }
```

I'd like for my users to get the `ProblemDetails` response either way. Also, I've been following this thread, and I too would like for exceptions not to be logged when returning `true`.
